android: react to change in network capabilities.

bznein opened the pull request:
NOTE: this is based on https://github.com/BitBoxSwiss/bitbox-wallet-app/pull/3421 so that one needs to be reviewed before this one. 


Use ConnectivityManager to react to onLost (connection entirely lost, e.g. going into airplane mode), onAvailable (the opposite), and onCapabilitiesChanged (other scenarios, for example the network loses access to the internet, but it still works locally)

Before asking for reviews, here is a check list of the most common things you might need to consider:
- [ ] updating the Changelog
- [ ] writing unit tests
- [ ] checking if your changes affect other coins or tokens in unintended ways
- [ ] testing on multiple environments (Qt, Android, ...)
- [ ] having an AI review your changes

benma commented on frontends/android/BitBoxApp/app/src/main/java/ch/shiftcrypto/bitboxapp/MainActivity.java:
```diff
@@ -467,6 +493,30 @@ public void run() {
+            @Override
+            public void onAvailable(android.net.Network network) {
+                super.onAvailable(network);
+                Mobileserver.setOnline(true);
```
onAvailable means there is a link (cable, WiFi), but not necessarily that you are online. E.g. WiFi catptive portals would return true here while not being connected. 

use 

```
NetworkCapabilities capabilities = connectivityManager.getNetworkCapabilities(network);
Mobileserver.setOnline(hasInternetConnectivity(capabilities));
```

benma commented on frontends/android/BitBoxApp/app/src/main/java/ch/shiftcrypto/bitboxapp/MainActivity.java:
```diff
@@ -85,6 +88,26 @@ public class MainActivity extends AppCompatActivity {
+            return true;
+        }
+
+        return capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET);
```
This is not enough

According to

https://developer.android.com/reference/android/net/NetworkCapabilities

> The NET_CAPABILITY_INTERNET capability means that the network is configured to reach the general Internet. It may or may not actually provide connectivity ; the NET_CAPABILITY_VALIDATED bit indicates that the system found actual connectivity to the general Internet the last time it checked. Apps interested in actual connectivity should usually look at both these capabilities. 

DeepSeek suggests:

```
    // Use VALIDATED where available (API 23+)
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        return capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED);
    }
    // Fallback for older devices
    return capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET);
```

benma reviewed (changes requested):
DeepSeek suggets these which seem reasonable:

---

3. **Improvements Needed**
   - Add `unregisterNetworkCallback()` in `onDestroy()` to prevent leaks
   - Handle null `activeNetwork` in connectivity checks
   - Use capability-aware `NetworkRequest`:
     ```java
     new NetworkRequest.Builder()
         .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
         .build()
     ```

bznein commented on frontends/android/BitBoxApp/app/src/main/java/ch/shiftcrypto/bitboxapp/MainActivity.java:
```diff
@@ -467,6 +499,30 @@ public void run() {
+            @Override
+            public void onAvailable(android.net.Network network) {
+                super.onAvailable(network);
+                checkConnectivity();
```
I think it's hard for them to be different, but I agree the current one seems better, will add a comment

bznein commented:
> DeepSeek suggets these which seem reasonable:
> 
> 3. **Improvements Needed**
>    
>    * Use capability-aware `NetworkRequest`:
>      ```java
>      new NetworkRequest.Builder()
>          .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
>          .build()
>      ```

I'm not sure about this one, the docs read:

>  Add the given capability requirement to this builder. These represent the requested network's required capabilities. Note that when searching for a network to satisfy a request, all capabilities requested must be satisfied.

My understanding is that if we add `NET_CAPABILITY_INTERNET`, the callback would not trigger when we are offline since this capability will not be satisfied by the current network. WDYT?

benma commented on frontends/android/BitBoxApp/app/src/main/java/ch/shiftcrypto/bitboxapp/MainActivity.java:
```diff
@@ -595,6 +666,7 @@ protected void onDestroy() {
             unbindService(connection);
         }
         super.onDestroy();
+        connectivityManager.unregisterNetworkCallback(networkCallback);
```
`if (connectivityManager != null && networkCallback != null) {` for robustness.

I also just noticed that the register/deregister pair does not match. onCreate pairs with onDestroy, and onStart pairs with onStop. but currently the stuff is registered in onStart. 

Would it make sense to just move the setup to onCreate instead?

benma commented on frontends/android/BitBoxApp/app/src/main/java/ch/shiftcrypto/bitboxapp/MainActivity.java:
```diff
@@ -595,6 +667,9 @@ protected void onDestroy() {
         }
         super.onDestroy();
+        if (connectivityManager != null && networkCallback != null) {
+            connectivityManager.unregisterNetworkCallback(networkCallback);
```
Register is in onStart, so unregister should be in onStop.

bznein commented:
@benma ready for review again, I tested it on Android and I believe it now covers all cases.

A few notes since many things changed since last version:

Removed callbacks to `onAvailable`, `onUnavailable`: these just return whether a network is available or not, no info about connectivity, and we should NOT call `getNetworkCapabilities` here, based on [the docs](https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback#onAvailable(android.net.Network):~:text=Do%20NOT%20call,to%20other%20callbacks.)

Similarly, changed `onLost` to NOT call the network capabilities changed callback. Based again on [the docs](https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback#onLost(android.net.Network):~:text=If%20the%20callback%20was%20registered%20with%20requestNetwork()%20or%20registerDefaultNetworkCallback()%2C%20it%20will%20only%20be%20invoked%20against%20the%20last%20network%20returned%20by%20onAvailable()%20when%20that%20network%20is%20lost%20and%20no%20other%20network%20satisfies%20the%20criteria%20of%20the%20request.) this will be called when the last network is lost, and no other network satisfies the criterias for the request

benma commented:
@bznein I fixed up my two nits myself as I am about to make a RC1 now, which will include this PR.

benma reviewed (approved):
tACK. There is a brief flashing of the banner when going from WiFi to mobile connection, but imho not a big deal.
